### PR TITLE
CTCTRADERS-990

### PR DESCRIPTION
**Pull request by MatthewAlexanderBeck**

### CODE REVIEW

File: app/controllers/ReferenceDataController.scala [hunk 1]
```diff
@@ -17,34 +17,53 @@
...
+            val refData: ReferenceDataPayload = ReferenceDataPayload(value.as[JsObject])
+
+            referenceDataService
+              .insert(refData)
+              .map {
+                case DataProcessingSuccessful => Accepted
+                case DataProcessingFailed     => InternalServerError(Json.toJsObject(ResponseErrorMessage(OtherError, None)))
+              }
+          case Left(error) => Future.successful(InternalServerError(Json.toJsObject(error)))
+        }
```
Comment (amaalali) [inline app/controllers/ReferenceDataController.scala:67]: The error handling looks good 👍 

Small suggestion, do we need getResult or would it be cleaner to put this in the yield?

File: app/services/GZipService.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+object GZipService {
+
+  def decompressArrayByteToJson(arrayByte: Array[Byte]): Either[ResponseErrorMessage, JsValue] =
+    Try {
+      val inputStream = new GZIPInputStream(new ByteArrayInputStream(arrayByte))
+      Json.parse(scala.io.Source.fromInputStream(inputStream).mkString)
+    } match {
+      case Success(value) => Right(value)
+      case Failure(_)     => Left(ResponseErrorMessage(OtherError, None))
+    }
```
Comment (amaalali) [inline app/services/GZipService.scala:40]: Really clean and easy. I also like that you're thinking about error handling here, looks good 👍

File: test/controllers/ReferenceDataControllerSpec.scala [hunk 3]
```diff
@@ -50,27 +54,94 @@ class ReferenceDataControllerSpec extends SpecBase with GuiceOneAppPerSuite with
...
+
+      status(result) mustBe Status.INTERNAL_SERVER_ERROR
+    }
+
+//    "returns Internal Server Error if request body is larger than memory threshold" in {
+//
+//      val result = route(app, fakeRequest(???)).value
+//
+//      status(result) mustBe Status.INTERNAL_SERVER_ERROR
+//    }
```
Comment (amaalali) [inline test/controllers/ReferenceDataControllerSpec.scala:144]: I don't think we should aim to test this in the unit tests, this tuning can be done with performance testing

**Review by amaalali — APPROVED** [review]
Looks good and clean!

**Review by amaalali — APPROVED** [review]

File: test/controllers/ReferenceDataControllerSpec.scala [hunk 4]
```diff
@@ -45,39 +49,62 @@ class ReferenceDataControllerSpec extends SpecBase with GuiceOneAppPerSuite with
...
-  private val testJson = Json.obj("foo" -> "bar")
+  def compress(input: Array[Byte]): Array[Byte] = {
+    val bos  = new ByteArrayOutputStream(input.length)
+    val gzip = new GZIPOutputStream(bos)
+    gzip.write(input)
+    gzip.close()
+    val compressed = bos.toByteArray
+    bos.close()
+    compressed
+  }
```
Comment (chris-ianson) [inline test/controllers/ReferenceDataControllerSpec.scala:60]: Is this used in 2 specs? Maybe consider putting it in spec base (looks like the same implementation used in both places)

**Review by chris-ianson — APPROVED** [review]
Looks good this @MatthewAlexanderBeck Minor comment in one of the specs but I've approved it

File: conf/app.routes [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # microservice specific routes
 
-GET   /resources  @controllers.ResourceLinksController.get
+GET     /resources                                  controllers.ResourceLinksController.get
 
 POST /reference-data-lists                          controllers.ReferenceDataController.referenceDataLists()
-POST /customs-office-lists/customs-office-lists     controllers.ReferenceDataController.customsOfficeLists()
```
Comment (amaalali) [inline conf/app.routes:6]: We can't change this as yet, this is what's in the spec, `/customs-office-lists/customs-office-lists`. We need to try to get this changed but for now that's what we need to implement

Comment (MatthewAlexanderBeck) [reply conf/app.routes:6]: Ah sorry, will revert this

**Pull request merged**